app/controllers/api/cloud_volumes_controller.rb

```diff
@@ -2,12 +2,37 @@ module Api
   class CloudVolumesController < BaseController
     include Subcollections::Tags
 
+    def create_resource(_type, _id = nil, data = {})
+      ext_management_system = ExtManagementSystem.find(data['ems_id'])
+
+      klass = CloudVolume.class_by_ems(ext_management_system)
+
+      validate = klass.validate_create_volume(ext_management_system)
+      raise validate[:message] unless validate[:available]
+
+      task_id = klass.create_volume_queue(session[:userid], ext_management_system, data)
+      action_result(true, "Creating Cloud Volume #{data['name']} for Provider: #{ext_management_system.name}", :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
     def delete_resource(type, id, _data = {})
       delete_action_handler do
         cloud_volume = resource_search(id, type, collection_class(:cloud_volumes))
         task_id = cloud_volume.delete_volume_queue(User.current_user)
         action_result(true, "Deleting Cloud Volume #{cloud_volume.name}", :task_id => task_id)
       end
     end
+
+    def options
+      return super unless params[:ems_id]
```

Comment:
Does this imply the user has to pass in the ems_id?  (Or do we know the ems_id based on the incoming endpoint?

That is, if a user does `OPTIONS /api/cloud_volumes/123` I would expect under the covers we would look up the ems_id and not force the user to enter it into the query.

Reply:
You can't really do `OPTIONS /api/cloud_volumes/123` if you want to create the very first cloud volume :confused: